### PR TITLE
drivers/servo: reimplement with high level interface

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The previous servo driver didn't provide any benefit over using PWM directly, as users controlled the servo in terms of PWM duty cycles. This changes the interface to provide a high level interface that abstracts the gory PWM details.

In addition, a SAUL layer and auto-initialization is provided.

### Testing procedure

The test application provides access to the servo driver via the `saul` shell command.

```
> saul
2022-08-02 22:12:31,826 # saul
2022-08-02 22:12:31,827 # ID	Class		Name
2022-08-02 22:12:31,830 # #0	ACT_SWITCH	LD1(green)
2022-08-02 22:12:31,832 # #1	ACT_SWITCH	LD2(blue)
2022-08-02 22:12:31,834 # #2	ACT_SWITCH	LD3(red)
2022-08-02 22:12:31,837 # #3	SENSE_BTN	B1(User button)
2022-08-02 22:12:31,838 # #4	ACT_SERVO	servo
> saul write 4 0
2022-08-02 22:12:41,443 # saul write 4 0
2022-08-02 22:12:41,445 # Writing to device #4 - servo
2022-08-02 22:12:41,447 # Data:	             0 
2022-08-02 22:12:41,450 # [servo] setting 0 to 2949 (0 / 255)
2022-08-02 22:12:41,453 # data successfully written to device #4
> saul write 4 256
2022-08-02 22:12:45,343 # saul write 4 256
2022-08-02 22:12:45,346 # Writing to device #4 - servo
2022-08-02 22:12:45,347 # Data:	           256 
2022-08-02 22:12:45,351 # [servo] setting 0 to 6865 (255 / 255)
2022-08-02 22:12:45,354 # data successfully written to device #4
```

Each write resulted in the MG90S servo that I connected to move to the corresponding position.

### Issues/PRs references

### CODE REVIEW

**Comment by maribu** [comment]
I updated the code to calculate the duty cycle at runtime, as the actual PWM frequency is not known on compile time. E.g. on the Arduino UNO the actual PWM frequency is 61 Hz (if I recall correctly). With that, the actual extension and the requested one differs by 22 % if using the nominal frequency instead of the actual one.

(Yes, this works fine on the Arduino UNO when rebased on top of https://github.com/RIOT-OS/RIOT/pull/18148 :))

File: drivers/include/servo.h [hunk 1]
```diff
@@ -11,82 +12,177 @@
...
+ *          "on duration"
+ * @param   pwm_freq        frequency of the PWM peripheral
+ * @param   pwm_res         resolution of the PWM peripheral
+ * @param   duration        duration of the "on phase" in microseconds
+ *
+ * @note    Scientific rounding is used
  */
-int servo_init(servo_t *dev, pwm_t pwm, int pwm_channel, unsigned int min, unsigned int max);
+#define SERVO_DUTY(pwm_freq, pwm_res, duration) \
+    ((1ULL * duration * (pwm_freq) * (pwm_res) + US_PER_SEC / 2) / US_PER_SEC)
```
Comment (benpicco) [inline drivers/include/servo.h:97]:
```suggestion
#define SERVO_DUTY(pwm_freq, pwm_res, duration) \
    DIV_ROUND(1ULL * (duration) * (pwm_freq) * (pwm_res), US_PER_SEC)
```

File: drivers/include/servo.h [hunk 2]
```diff
@@ -11,82 +12,177 @@
...
+ * @brief   PWM configuration parameters as **XFA constant array**
  *
- * In case pos is larger/smaller then the max/min values, pos will be set to
- * these values.
+ * @note    Use @ref XFA_CONST to extend this array
+ */
+extern const servo_pwm_params_t servo_pwm_params[];
+
+/**
+ * @brief   SAUL entires for the servos as **XFA array**
```
Comment (benpicco) [inline drivers/include/servo.h:147]:
```suggestion
 * @brief   SAUL entries for the servos as **XFA array**
```

File: drivers/include/servo.h [hunk 3]
```diff
@@ -11,82 +12,177 @@
...
+ *
+ * @note    Use @ref XFA to extend this array
+ * @details The user does not need to initialize the entires, as this is done
```
Comment (benpicco) [inline drivers/include/servo.h:150]:
```suggestion
 * @details The user does not need to initialize the entries, as this is done
```

File: drivers/servo/servo.c [hunk 4]
```diff
@@ -16,92 +17,97 @@
...
+ * @param   pwm_freq        frequency of the PWM peripheral
+ * @param   pwm_res         resolution of the PWM peripheral
+ * @param   duration        duration of the "on phase" in microseconds
+ *
+ * @note    Scientific rounding is used
+ */
+static uint16_t duty_cycle(uint32_t freq, uint16_t res, uint16_t duration_us)
 {
-    int actual_frequency;
+    return ((uint64_t)duration_us * freq * res + US_PER_SEC / 2) / US_PER_SEC;
```
Comment (benpicco) [inline drivers/servo/servo.c:51]:
```suggestion
    return DIV_ROUND((uint64_t)duration_us * freq * res, US_PER_SEC);
```

**Review by benpicco — COMMENTED** [review]
Yea that's a much better API.
I don't quite understand the reason for use of XFA - why not use the same pattern as the other drivers?

**Comment by maribu** [comment]
> I don't quite understand the reason for use of XFA - why not use the same pattern as the other drivers?

Likely I should have done the conversion as a follow up. Note that there are two things that are changed: First, it allocates device states and params as an arrays and uses an index has handle instead of a pointer. Second, it uses XFA to implement the arrays.

The use of arrays has the following advantages:
- It safes RAM, as the current pattern of copying the params stuff into the state is no longer needed. (The alternative would be to add a pointer from the state to params, which would increase the state size by one pointer and add one level of pointer traversal to accessing params.)
- And most importantly: It fixes the fundamentally broken sensor and actuator auto initialization.
    - Auto-initialization does happen if used with SAUL, but then there is no way to access the device state and use it without the SAUL API.
    - When not using SAUL, currently no way of auto-initialization is possible
    - Using indexes solves the issue of obtaining the device pointer when initialization happens in auto_init

The use of XFA has the advantage that a user can easily add additional devices. E.g. let's say board foo already has 3 LEDs exposed via SAUL. Now we add additional LEDs via a shield. With XFA, we can just extend the LED array without having to mess with the board LED config.

Likely I should update an existing driver, maybe the SHT1x, to the XFA approach, first.

**Comment by maribu** [comment]
I dropped the use an of index-based `servo_t` and of XFA for consistency. The `servo` shell command had to be removed as well, as co-existence with SAUL no longer is possible.

This can be added in a follow up. Let's fix the servo API first.

**Review by benpicco — COMMENTED** [review]
CI has some comments, otherwise looks good

**Review by benpicco — APPROVED** [review]

**Comment by maribu** [comment]
I greatly changed this PR, as I had urgent need to use this with the nRF52840. That MCU however cannot create PWM anywhere near the 50 Hz to 100 Hz required (I think the biggest prescaler could device the clock down to something like 32.678 or so). So I made `servo_pwm` and `servo_timer` submodules that contain different implementations of the same user facing API. The `servo_timer` implementation uses `periph_timer_periodic` to basically implement a 50 Hz PWM signal in software. Hence, `servo_pwm` remains the preferred implementation for every board that can generate slow enough PWM signals.

I'll dismiss the review now in the light of the mayor changes I just pushed.

**Review by benpicco — COMMENTED** [review]
Sorry, I completely forgot about this one.
The pwm timer might need some explanation wrt multiple channels, but otherwise looks good.

File: drivers/include/servo.h [hunk 5]
```diff
@@ -11,82 +12,210 @@
...
+#include <stdint.h>
+
 #include "periph/pwm.h"
+#include "periph/timer.h"
+#include "saul.h"
+#include "saul_reg.h"
+#include "time_units.h"
+
+#ifndef SERVO_TIMER_MAX_CHAN
+#define SERVO_TIMER_MAX_CHAN    4
```
Comment (benpicco) [inline drivers/include/servo.h:55]: doccheck is right, what does this mean? 
Multiple servos per timer?

**Review by benpicco — APPROVED** [review]
Please squash

**Comment by maribu** [comment]
I forgot the KConfig integration and added it now.

**Comment by benpicco** [comment]
> I forgot the KConfig integration and added it now.

pretty much impossible to get that right on the first try :wink:

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
smpboot_control

**Comment by benpicco** [comment]
bors cancel

**Comment by benpicco** [comment]
bors merge

**Comment by maribu** [comment]
bors retry

Now with a `Makefile.ci`, let's see what happens now.

**Comment by maribu** [comment]
bors retry

**Comment by maribu** [comment]
Thx :)

**Pull request merged**